File: week10/19/P2.py
class Solution(object):
    def stoneGameIX(self, stones):
        stoneDict = {0:0, 1:0, 2:0}
        for s in stones: # 0,1,2 dict으로 전환
            num = s % 3
            stoneDict[num] += 1
        if (stoneDict[0] % 2 == 0): # 0이 짝수일때
            if (stoneDict[1] > 0 and stoneDict[2] > 0): # 선턴 잡으면 끝
                return True
            else:
                return False
        else: # 0이 홀수일때
            if (abs(stoneDict[1] - stoneDict[2]) > 2): # 턴이 한번 넘어가므로 2개 넘어가도록 차이가 발생해야 함
                return True
            else:
                return False

if __name__ == "__main__":
    result = Solution()
    stones = [2]
    print(result.stoneGameIX(stones)) # False
    stones = [2,1]
    print(result.stoneGameIX(stones)) # True
    stones = [2,2]
    print(result.stoneGameIX(stones)) # False
    stones = [0]
    print(result.stoneGameIX(stones)) # False
    stones = [2,0]
    print(result.stoneGameIX(stones)) # False
    stones = [2,1,0]
    print(result.stoneGameIX(stones)) # False
    stones = [2,2,0]
    print(result.stoneGameIX(stones)) # False
    stones = [2,2,2,0]
    print(result.stoneGameIX(stones)) # True

File: week10/19/P2.c
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>

bool stoneGameIX(int* stones, int stonesSize) {
    int stoneCount[3] = {0, 0, 0}; // {0, 1, 2} 모듈로 결과에 따라 돌을 세는 배열

    for (int i = 0; i < stonesSize; i++) {
        int num = stones[i] % 3;
        stoneCount[num]++;
    }

    if (stoneCount[0] % 2 == 0) {
        if (stoneCount[1] > 0 && stoneCount[2] > 0) return true;
        else return false;
    } else {
        if (abs(stoneCount[1] - stoneCount[2]) > 2) return true;
        else return false;
    }
}

int main() {
    int stones[] = {2,2,2,0};
    printf("%d\n", stoneGameIX(stones, 4));
    
    return 0;
}